src/device/socket/multiconnectionmanager.rs

```diff
@@ -0,0 +1,763 @@
+    /// shutdown.
+    pub fn shutdown(&mut self, destination: VsockAddr, src_port: u32) -> Result {
+        let connection = self
+            .connections
+            .iter()
+            .find(|connection| {
+                connection.info.dst == destination && connection.info.src_port == src_port
+            })
+            .ok_or(SocketError::NotConnected)?;
+
+        self.driver.shutdown(&connection.info)
+    }
+
+    /// Forcibly closes the connection without waiting for the peer.
+    pub fn force_close(&mut self, destination: VsockAddr, src_port: u32) -> Result {
+        let (index, connection) = self
+            .connections
+            .iter()
+            .enumerate()
+            .find(|(_, connection)| {
+                connection.info.dst == destination && connection.info.src_port == src_port
+            })
+            .ok_or(SocketError::NotConnected)?;
+
+        self.driver.force_close(&connection.info)?;
+
+        self.connections.swap_remove(index);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct RingBuffer {
+    buffer: Box<[u8]>,
+    /// The number of bytes currently in the buffer.
+    used: usize,
+    /// The index of the first used byte in the buffer.
+    start: usize,
+}
+
+impl RingBuffer {
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            buffer: FromBytes::new_box_slice_zeroed(capacity),
+            used: 0,
+            start: 0,
+        }
+    }
+
+    /// Returns the number of bytes currently used in the buffer.
+    pub fn used(&self) -> usize {
+        self.used
+    }
+
+    /// Returns true iff there are currently no bytes in the buffer.
+    pub fn is_empty(&self) -> bool {
+        self.used == 0
+    }
+
+    /// Returns the number of bytes currently free in the buffer.
+    pub fn available(&self) -> usize {
+        self.buffer.len() - self.used
+    }
+
+    /// Adds the given bytes to the buffer if there is enough capacity for them all.
+    ///
+    /// Returns true if they were added, or false if they were not.
+    pub fn write(&mut self, bytes: &[u8]) -> bool {
+        if bytes.len() > self.available() {
+            return false;
+        }
+
+        let end = (self.start + self.used) % self.buffer.len();
+        let write_before_wraparound = min(bytes.len(), self.buffer.len() - end);
+        let write_after_wraparound = bytes
+            .len()
+            .checked_sub(write_before_wraparound)
+            .unwrap_or_default();
+        self.buffer[end..end + write_before_wraparound]
+            .copy_from_slice(&bytes[0..write_before_wraparound]);
+        self.buffer[0..write_after_wraparound].copy_from_slice(&bytes[write_before_wraparound..]);
+        self.used += bytes.len();
+
+        true
+    }
+
+    /// Reads and removes as many bytes as possible from the buffer, up to the length of the given
+    /// buffer.
+    pub fn read(&mut self, out: &mut [u8]) -> usize {
```

Comment:
Since the buffer will be dropped after reading, what about renaming this to `pop` and the `write` function `push`?

Reply:
It's true this is more of a queue, I think `add` is okay but I don't see `drain` so much used for queues. What about `add` and `remove` like the Queue API in java?